internal/gitfs/gitfs.go

```diff
@@ -0,0 +1,349 @@
+package gitfs
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"time"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/filemode"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/storage"
+	"go.flipt.io/flipt/internal/containers"
+	"go.uber.org/zap"
+)
+
+// FS is a filesystem implementation that decorates a git storage
+// implementation and instance of a git tree.
+// FS implements fs.FS and is therefore read-only by design.
+type FS struct {
+	logger *zap.Logger
+
+	storage storage.Storer
+	tree    *object.Tree
+}
+
+// New instantiates and instance of storage which retrieves files from
+// the provided storage.Storer implementation and instance of *object.Tree.
+func New(storage storage.Storer, tree *object.Tree) FS {
+	return FS{zap.NewNop(), storage, tree}
+}
+
+// Options configures call to NewFromRepo.
+type Options struct {
+	ref plumbing.ReferenceName
+}
+
+// WithReference overrides the default reference to main.
+func WithReference(ref plumbing.ReferenceName) containers.Option[Options] {
+	return func(o *Options) {
+		o.ref = ref
+	}
+}
+
+// NewFromRepo is a convience utility which constructs an isntance of FS
+// from the provided git repository.
+// By default the returned FS serves the content from the root tree
+// for the commit at reference HEAD.
+func NewFromRepo(repo *git.Repository, opts ...containers.Option[Options]) (FS, error) {
+	o := Options{ref: plumbing.HEAD}
+	containers.ApplyAll(&o, opts...)
+
+	ref, err := repo.Reference(o.ref, true)
+	if err != nil {
+		return FS{}, fmt.Errorf("resolving reference (%q): %w", o.ref, err)
+	}
+
+	return NewFromRepoHash(repo, ref.Hash())
+}
+
+// NewFromRepoHash is a convience utility which constructs an isntance of FS
+// from the provided git repository and hash.
+func NewFromRepoHash(repo *git.Repository, hash plumbing.Hash) (FS, error) {
+	commit, err := repo.CommitObject(hash)
+	if err != nil {
+		return FS{}, fmt.Errorf("fetching commit (%q): %w", hash, err)
+	}
+
+	tree, err := repo.TreeObject(commit.TreeHash)
+	if err != nil {
+		return FS{}, fmt.Errorf("retrieving root tree (%q): %w", commit.TreeHash, err)
+	}
+
+	return New(repo.Storer, tree), nil
+}
+
+// Open opens the named file.
+// When Open returns an error, it should be of type *PathError
+// with the Op field set to "open", the Path field set to name,
+// and the Err field describing the problem.
+//
+// Open should reject attempts to open names that do not satisfy
+// ValidPath(name), returning a *PathError with Err set to
+// ErrInvalid or ErrNotExist.
+func (f FS) Open(name string) (_ fs.File, err error) {
+	defer func() {
+		if err != nil {
+			err = opPathError("Open", name)(err)
+		}
+	}()
+
+	if !fs.ValidPath(name) {
+		return nil, fs.ErrInvalid
+	}
```

Comment:
Just clarifying my understanding here, this picks up the deferred error wrapping because of the named return values?

Reply:
Thats correct 👍 The defer runs after `return` at which point `err` has been populated by whatever the `return` statement has in that argument position.

This is a lazy way of wrapping all things 😂 :
https://go.dev/play/p/lG4C8IUiW7C

Reply:
@GeorgeMac Makes sense. Thanks!